### ISSUE TITLE
Raft shutdown doesn't wait for replication streams to shut down

### ISSUE THREAD

**Issue by lilyball** [railgun-rs/actix-raft#138]
When Raft is told to shutdown, it waits for its core to shut down. If it was in the leader state, it signals the replication streams to shut down too. But it doesn't actually wait for the replication streams to shut down. The effect of this is Raft can declare that it has shut down and return from the shutdown future while the replication streams are still alive and talking to other nodes.

Raft really needs to ensure all of its operations have shut down before it returns from the shutdown future.

**Comment by drmingdrmer** [comment]
AFAIK, when the LeaderState is dropped, the channel for communicating to the replication stream will be dropped too.
And then replication task will find it is trying to `recv` from a closed channel then shut itself down.

There are several checking code as the following:
https://github.com/async-raft/async-raft/blob/bcc246ccf8fa2858baf8e07bf760ee6db4d85cf6/async-raft/src/replication/mod.rs#L550-L556

Thus the replication task may keep running for a little while before draining its buffer.

**Comment by lilyball** [comment]
> Thus the replication task may keep running for a little while before draining its buffer.

Right, and this is the surprising part to me. When I ask the library to shut down, it gives me a future. Thus, when the future resolves, I expect this to mean that the library has shut down. But the replication streams may still be operating (and I'm worried there may be other orphaned futures running like this too).

The problem isn't that the replication streams don't shut down immediately, it's that async-raft claims that Raft has shutdown when it hasn't actually stopped everything.

One potential issue here is if I shut down Raft and then immediately restart it, my process may now be attempting to replicate to a remote node while also doing whatever else it does in this scenario (I'm not sure offhand; does it try to run leader election, or does it discover that everyone else still thinks it's the leader and thus start up another set of replication streams, or something else? Whatever it does, it's likely inconsistent with having existing zombie replication streams).

**Comment by lilyball** [comment]
Also note that `LineRateState` at the least always drains its outbound buffers prior to checking the channel. So if it's got a bunch of stuff buffered (or I/O is particularly slow) the stream could take an arbitrarily long time before shutting down. See #136 for more details on this part.

**Comment by drmingdrmer** [comment]
> One potential issue here is if I shut down Raft and then immediately restart it, my process may now be attempting to replicate to a remote node while also doing whatever else it does in this scenario (I'm not sure offhand; does it try to run leader election, or does it discover that everyone else still thinks it's the leader and thus start up another set of replication streams, or something else? Whatever it does, it's likely inconsistent with having existing zombie replication streams).

You do not need to worry about the stop-restart scenario. At least it won't affect the correctness.

A new leader always runs with higher term. There may be orphan replication task still running. Once a follower received a request from the new leader, it will discard any other request with lower term.

It is possible that a new leader continue using the previous term and the correctness still holds. But raft does not do it this way. It always elects itself first.
